tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
 package langserver
 
+import (
+	"core"
+	"parse"
+	"strings"
+
+	"parse/asp"
+	"tools/build_langserver/lsp"
+	"io/ioutil"
+	"fmt"
+)
+
 // TODO(bnmetrics): This file should contain functions to retrieve builtin and custom definitions of build defs
+
+// Analyzer is a wrapper around asp.Analyzer
+// This is being loaded into a handler on initialization
+type Analyzer struct {
+	parser *asp.Parser
+	BuiltIns map[string]*RuleDef
+}
+
+// RuleDef is a wrapper around asp.FuncDef,
+// it also includes a Header(function definition)
+// And Argument map stores the name and the information of the arguments this rule has
+type RuleDef struct {
+	*asp.FuncDef
+	Header string
+	ArgMap map[string]*Argument
+}
+
+// Argument is a wrapper around asp.Argument,
+// this is used to store the argument information for specific rules,
+// and it also tells you if the argument is required
+type Argument struct {
+	*asp.Argument
+	definition string
+	required bool
+}
+
+// Identifier is a wrapper around asp.Identifier
+// Including the starting line and the ending line number
+type Identifier struct {
+	*asp.IdentStatement
+	StartLine int
+	EndLine int
+}
+
+func newAnalyzer() *Analyzer {
+	state := core.NewDefaultBuildState()
+	parser := parse.GetParserWithBuiltins(state)
+
+	a := &Analyzer{
+		parser:parser,
+	}
+	a.BuiltInsRules()
+
+	return a
+}
+
+// BuiltInsRules gets all the builtin functions and rules as a map, and store it in Analyzer.BuiltIns
+func (a *Analyzer) BuiltInsRules() {
+	statementMap := make(map[string]*RuleDef)
+
+	for _, statement := range a.parser.GetAllBuiltinStatements() {
+		// Saves FuncDefs into the statementMap if it's a None private rule
+		if statement.FuncDef != nil && !strings.HasPrefix(statement.FuncDef.Name, "_") {
```

Comment:
I wonder if we should promote the underscore-private business to a more official concept (like maybe add a bool to the FuncDef or something). Thoughts?

Reply:
Yes, I think I like the idea of prefix `_` for private one, and at the same time funcDef could probably have an attribute called `IsPrivate`. If a function does have prefix `_`, we'd have `IsPrivate` as true. What do you think?

Reply:
Yup exactly. Should save checking for the same idiom in multiple places.